test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_image_arguments.cpp

```diff
@@ -0,0 +1,416 @@
+        test_error(error, "clEnqueueCommandBufferKHR failed")
+
+            size_t origin[3] = { 0, 0, 0 };
+        size_t region[3] = { image_desc.image_width, 1, 1 };
+
+        error = clEnqueueReadImage(queue, new_image, CL_TRUE, origin, region, 0,
+                                   0, outputData, 0, nullptr, nullptr);
+        test_error(error, "clEnqueueReadImage failed");
+
+        for (size_t i = 0; i < imageInfo.width; ++i)
+        {
+            if (imageValues[i] != outputData[i])
+            {
+                log_error("Data failed to verify: imageValues[%d]=%d != "
+                          "outputData[%d]=%d\n",
+                          i, imageValues[i], i, outputData[i]);
+
+                return TEST_FAIL;
+            }
+        }
+
+        return TEST_PASS;
+    }
+
+    cl_mutable_command_khr command = nullptr;
+};
+
+struct MutableDispatchImage2DArguments : public BasicMutableCommandBufferTest
+{
+    using BasicMutableCommandBufferTest::BasicMutableCommandBufferTest;
+
+    MutableDispatchImage2DArguments(cl_device_id device, cl_context context,
+                                    cl_command_queue queue)
+        : BasicMutableCommandBufferTest(device, context, queue)
+    {}
+
+    virtual cl_int SetUp(int elements) override
+    {
+        BasicMutableCommandBufferTest::SetUp(elements);
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+
+        const char *sample_const_arg_kernel =
+            R"(__kernel void sample_test( read_only image2d_t source, sampler_t
+            sampler, int width, write_only image2d_t dest, __global int4
+        *results )
+            {
+               int x = get_global_id(0);
+               int y = get_global_id(1);
+
+               int4 color = read_imagei( source, sampler, (int2) (x, y) );
+
+               write_imagei( dest, (int2) (x, y), color );
+            })";
+
+        cl_int error;
+        clProgramWrapper program;
+        clKernelWrapper kernel;
+
+        cl_image_desc image_desc;
+        memset(&image_desc, 0x0, sizeof(cl_image_desc));
+        image_desc.image_type = CL_MEM_OBJECT_IMAGE2D;
+        image_desc.image_width = 4;
+        image_desc.image_height = 4;
+        image_desc.image_row_pitch = 0;
+        image_desc.num_mip_levels = 0;
+
+        size_t data_size =
+            image_desc.image_width * image_desc.image_height * sizeof(cl_int);
+
+        const cl_image_format formats = { CL_RGBA, CL_UNSIGNED_INT8 };
+
+        image_descriptor imageInfo = { 0 };
+        imageInfo.type = CL_MEM_OBJECT_IMAGE2D;
+        imageInfo.width = 4;
+        imageInfo.height = 4;
+        imageInfo.format = &formats;
+
+        BufferOwningPtr<char> imageValues;
+
+        MTdataHolder d(gRandomSeed);
+        generate_random_image_data(&imageInfo, imageValues, d);
+
+        char *host_ptr = (char *)imageValues;
+        BufferOwningPtr<char> outputData(malloc(data_size));
```

Comment:
can we just use a `std::vector` here rather than `BufferOwningPtr`?

Reply:
It's not compatible with random number generator.

Reply:
I don't see `outputData` being passed into a random number generator in this test, only the 1D test

Reply:
We are comparing it against `imageValues` when checking the results.

Reply:
your indexing into both `outputData` and `imageValues` with `operator[]` so we're comparing `char`s when checking the results not the containers themselves.

When I change this locally the test runs and passes fine:
```diff
diff --git a/test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_image_arguments.cpp b/test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_image_arguments.cpp
index d770a75..9803a8a 100644
--- a/test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_image_arguments.cpp
+++ b/test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_image_arguments.cpp
@@ -272,7 +272,7 @@ struct MutableDispatchImage2DArguments : public BasicMutableCommandBufferTest
         generate_random_image_data(&imageInfo, imageValues, d);

         char *host_ptr = (char *)imageValues;
-        BufferOwningPtr<char> outputData(malloc(data_size));
+        std::vector<char> outputData(data_size);

         clMemWrapper src_image =
             create_image_2d(context, CL_MEM_READ_WRITE | CL_MEM_COPY_HOST_PTR,
@@ -377,7 +377,7 @@ struct MutableDispatchImage2DArguments : public BasicMutableCommandBufferTest
                              1 };

         error = clEnqueueReadImage(queue, new_image, CL_TRUE, origin, region, 0,
-                                   0, outputData, 0, nullptr, nullptr);
+                                   0, outputData.data(), 0, nullptr, nullptr);
         test_error(error, "clEnqueueReadImage failed");

         for (size_t i = 0; i < imageInfo.width * imageInfo.height; ++i)
```